Add local build for System.Globalization shim

smhmhmd opened the pull request:
Bundle the System.Globalization.Native shim sources into the native AOT toolchain package and compile them automatically using C/C++ compiler on the target machine.

Compiling of the shim sources using C/C++ compiler is done during dotnet publish as follows:
  dotnet publish  /p:LocalSystemGlobalizationNative=1 -bl -r linux-x64

This change is 'opt-in', it is enabled only by using /p:LocalSystemGlobalizationNative=1

Commit:
Add local build for System.Globalization shim

Bundle the System.Globalization.Native shim sources into the native AOT toolchain package and compile them automatically using C/C++ compiler on the target machine.

Compiling of the shim sources using C/C++ compiler is done during dotnet publish as follows:
  dotnet publish  /p:LocalSystemGlobalizationNative=1 -bl -r linux-x64

This change is 'opt-in', it is enabled only by using /p:LocalSystemGlobalizationNative=1

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=72896) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=72896) before we can accept your contribution.<br/><br/>:x: smhmhmd [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=72896)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=72896) it.</sub>

smhmhmd commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:
```diff
@@ -64,7 +64,10 @@ The .NET Foundation licenses this file to you under the MIT license.
       <NetCoreAppNativeLibrary Include="@(NetCoreAppNativeLibrary->'%(Identity)')">
         <EscapedPath>$(IlcFrameworkNativePath)lib%(Identity).a</EscapedPath>
       </NetCoreAppNativeLibrary>
+      <NativeLibrary Include="$(NugetPackageRoot)/microsoft.dotnet.ilcompiler/7.0.0-dev/shim_source_code/libs/System.Globalization.Native/build/libSystem.Globalization.Native-Static.a" Condition="'$(LocalSystemGlobalizationNative)' != ''"/>
```
@jkotas  I need your help to replace the hardcoded "7.0.0-dev"

jkotas replied:
NugetPackageRoot is nuget cache. The nuget tool is the only one allowed to write into nuget cache.

This should get compiled to `IntermediateOutputPath` instead. It will also fix the hardcoded `7.0.0-dev`.

smhmhmd replied:
Addressed in next revision using IntermediateOutputPath

tarekgh commented:
CC @jkotas

tarekgh commented:
https://github.com/dotnet/runtime/issues/70848

jkotas commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:
```diff
@@ -64,7 +64,10 @@ The .NET Foundation licenses this file to you under the MIT license.
       </NetCoreAppNativeLibrary>
+      <NativeLibrary Include="$(NugetPackageRoot)/microsoft.dotnet.ilcompiler/7.0.0-dev/shim_source_code/libs/System.Globalization.Native/build/libSystem.Globalization.Native-Static.a" Condition="'$(LocalSystemGlobalizationNative)' != ''"/>
       <NativeLibrary Include="@(NetCoreAppNativeLibrary->'%(EscapedPath)')" />
+      <StaticICULibs Include="/usr/lib/x86_64-linux-gnu/libicuio.a;/usr/lib/x86_64-linux-gnu/libicutu.a;/usr/lib/x86_64-linux-gnu/libicui18n.a;/usr/lib/x86_64-linux-gnu/libicuuc.a;/usr/lib/x86_64-linux-gnu/libicudata.a" Condition="'$(LocalSystemGlobalizationNative)' != '' And Exists('/usr/lib/x86_64-linux-gnu/libicuio.a')" />
```
These hardcoded paths with x86_64 need to be fixed too.

smhmhmd replied:
Addressed in next revision

jkotas commented on src/native/libs/System.Globalization.Native/pal_icushim_native_static.c:
```diff
@@ -0,0 +1,150 @@
+#include <unicode/ucurr.h>
```
This should not be needed. The existing `STATIC_ICU` has the same effect - just need to define it when building.

smhmhmd replied:
Addressed in next revision

eerhardt commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:
```diff
@@ -74,8 +77,11 @@ The .NET Foundation licenses this file to you under the MIT license.
       <NativeFramework Include="GSS" />
     </ItemGroup>
 
+    <Exec Command="if [ -d $(NugetPackageRoot)/microsoft.dotnet.ilcompiler/7.0.0-dev/shim_source_code ]; then cd $(NugetPackageRoot)/microsoft.dotnet.ilcompiler/7.0.0-dev/shim_source_code/libs/System.Globalization.Native/  %26%26  cp CMakeLists-standalone.txt CMakeLists.txt %26%26 cp config-standalone.h config.h %26%26 mkdir -p build %26%26 cd build %26%26 cmake ../ %26%26 make -j; fi" Condition= ...
```
Would it be more maintainable if this shell code was in its own `.sh` file?

smhmhmd replied:
Add separate script in next revision

eerhardt commented on src/native/libs/System.Globalization.Native/pal_icushim_native_static.c:
```diff
@@ -0,0 +1,150 @@
+#include <unicode/ucurr.h>
```
Does this need a copyright header?

smhmhmd replied:
Added in next revision

eerhardt commented on src/native/libs/System.Globalization.Native/config-standalone.h:
```diff
@@ -0,0 +1,4 @@
+#pragma once
```
copyright header?

smhmhmd replied:
Added in next revision

am11 commented:
So far we have not required cmake dependency for `PublishAot=tru` on the user machine, I think it would be nice to keep it that way.

jkotas commented:
> So far we have not required cmake dependency for PublishAot=tru on the user machine, I think it would be nice to keep it that way

I do not think that cmake dependency is a problem for this special mode of NativeAOT publishing. This mode can grow into optionally building all C/C++ that the runtime is composed from.

am11 commented:
We can construct the same arguments in out existing msbuild targets which invokes the compiler, without requiring a fully fledged build system from user. It just feels heavy (even for this optional mode) as there is no system introspection going on only a few paths and macro definitions which can be passed with rest of the command line arguments.

smhmhmd committed:
Merge branch 'dotnet:main' into build

smhmhmd committed:
Incorporate review feedback

* Copy shim source code into users Obj directory at IntermediateOutputPath and compile
** $ ls -al obj/Debug/net7.0/linux-x64/shim_source_code/libs/System.Globalization.Native/build/libSystem.Globalization.Native-Static.a

* Add shell script for readability

* Add copyright headers

smhmhmd committed:
Fix IlCompilerPackagePath

Add change that was part of previous commit

jkotas commented:
> > This should not be needed. The existing STATIC_ICU has the same effect - just need to define it when building.

> Addressed in next revision

This does not seem to be addressed. `pal_icushim_native_static.c` file is still there.

jkotas commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:
```diff
@@ -74,8 +76,16 @@ The .NET Foundation licenses this file to you under the MIT license.
       <NativeFramework Include="GSS" />
     </ItemGroup>
 
+    <ItemGroup>
```
`<IlcPath>` should point to the package root. Can you use that instead of trying to guess it here?

smhmhmd replied:
I used <IlcSdkPath> in next revision

smhmhmd replied:
I hope using `IlcSdkPath` is good

jkotas replied:
Was there a problem with `IlcPath`? It would make the `..` go away.

smhmhmd committed:
Address review feedback

* Include existing pal_icushim_static.c by using STATIC_ICU define
* Use IlcSdkPath to copy shim source code

jkotas commented on src/native/libs/System.Globalization.Native/pal_icushim.c:
```diff
@@ -481,6 +481,10 @@ int32_t GlobalizationNative_LoadICU()
     FOR_ALL_ICU_FUNCTIONS
     ValidateICUDataCanLoad();
 
+#if defined(USE_ICUSHIM_STATIC_BUILD)
```
Is this needed?

smhmhmd replied:
Yes, the issue is that the entrypoint is different, dotnet uses `GlobalizationNative_LoadICU` instead of `GlobalizationNative_LoadICUData`

jkotas replied:
I am not able to find the place where `USE_ICUSHIM_STATIC_BUILD` is defined. Where is it defined?

smhmhmd replied:
I changed it to STATIC_SHIM_COMPILE in latest revision, it is in CMakeLists-standalone.txt

smhmhmd replied:
You are right, I need to remove it from line 484 as well, thanks to @Beau-Gosse-dev for pointing out.

jkotas commented on src/installer/pkg/projects/Microsoft.DotNet.ILCompiler/Microsoft.DotNet.ILCompiler.pkgproj:
```diff
@@ -21,6 +21,9 @@
       <File Include="$(CoreCLRBuildIntegrationDir)*" TargetPath="build" />
       <File Include="$(CoreCLRILCompilerDir)netstandard\*" TargetPath="tools/netstandard" />
       <File Include="sdk\Sdk.props" TargetPath="Sdk" />
+      <File Include="$(MSBuildThisFileDirectory)\..\..\..\..\native\libs\System.Globalization.Native\*" TargetPath="shim_source_code/libs/System.Globalization.Native"/>
```
I would call it `source_code` or just `src`. I expect we may end up putting more than just the shims there over time.

smhmhmd replied:
ok, will rename it to `source_code`

jkotas commented on src/native/libs/System.Globalization.Native/config-standalone.h:
```diff
@@ -0,0 +1,8 @@
+
+#pragma once
+
+#define HAVE_UDAT_STANDALONE_SHORTER_WEEKDAYS 1
```
Should we do regular config for this instead of hardcoding?

smhmhmd replied:
ok, will do regular config

smhmhmd commented:
> > > This should not be needed. The existing STATIC_ICU has the same effect - just need to define it when building.
> 
> > Addressed in next revision
> 
> This does not seem to be addressed. `pal_icushim_native_static.c` file is still there.

Removed `pal_icushim_native_static.c` in latest revision

smhmhmd committed:
Incorporate review feedback

* Remove USE_ICUSHIM_STATIC_BUILD in pal_icushim.c
* Rename shim_source_code to source_code in Microsoft.DotNet.ILCompiler.pkgproj and script
* Regular config is used, config-standalone.h is removed
* IlcPath/.. is replaced to make the ".." go away

Beau-Gosse-dev commented:
For what it's worth, I tested this on ARM and it worked the same as x64 as of commit [6075e03](https://github.com/dotnet/runtime/pull/72896/commits/6075e0316bb34a1f0629a18b2cb3481a9ba2de77)

jkotas commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:
```diff
@@ -74,8 +76,16 @@ The .NET Foundation licenses this file to you under the MIT license.
       <NativeFramework Include="GSS" />
     </ItemGroup>
 
+    <Message Text="IlcPath = $(IlcPath)" />
```
You can get this path from the verbose log. Printing it all the time adds unnecessary clutter to the default output.

smhmhmd replied:
Ok, will remove the prints.

jkotas commented on src/native/libs/System.Globalization.Native/build-nativeaot-shim-standalone.sh:
```diff
@@ -0,0 +1,34 @@
+INTERMEDIATE_OUTPUT_PATH=$2
+
+if [ -d "$SHIM_SOURCE_DIR" ]; then
+    cp -r "$SHIM_SOURCE_DIR" "$INTERMEDIATE_OUTPUT_PATH"
```
Why do you need to copy all sources over? Can you just point the build at the sources in the package and only set the output the intermediate path?

jkotas commented on src/native/libs/System.Globalization.Native/CMakeLists-standalone.txt:
```diff
@@ -0,0 +1,45 @@
+add_compile_options(-Wno-extra-semi-stmt)
+add_compile_options(-Wno-unknown-warning-option)
+
+add_definitions(-D_GNU_SOURCE -g -fPIC -DBUILDENV_DEBUG=1 -DDEBUG -DDISABLE_CONTRACTS -DHOST_64BIT -DHOST_AMD64 -DHOST_UNIX -DPALEXPORT=EXTERN_C -DTARGET_64BIT -DTARGET_AMD64 -DTARGET_LINUX -DTARGET_UNIX -DURTBLDENV_FRIENDLY=Debug -D_DBG -D_DEBUG -D_FILE_OFFSET_BITS=64 -DSTATIC_ICU -DSTATIC_SHIM_COMPILE)
```
This list of defines needs pruning to just what's actually needed. For example `BUILDENV_DEBUG`, `DEBUG` or `TARGET_AMD64` should not be needed (there are probably more).

jkotas commented on src/native/libs/System.Globalization.Native/CMakeLists-standalone.txt:
```diff
@@ -0,0 +1,45 @@
+
+include(configure.cmake)
+
+if(CMAKE_SYSTEM_NAME STREQUAL "Linux") # Only Linux for now
```
We should print error when this is run on non-Linux.

jkotas commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:
```diff
@@ -64,7 +64,9 @@ The .NET Foundation licenses this file to you under the MIT license.
       <NetCoreAppNativeLibrary Include="@(NetCoreAppNativeLibrary->'%(Identity)')">
         <EscapedPath>$(IlcFrameworkNativePath)lib%(Identity).a</EscapedPath>
       </NetCoreAppNativeLibrary>
+      <NativeLibrary Include="$(IntermediateOutputPath)/source_code/libs/System.Globalization.Native/build/libSystem.Globalization.Native-Static.a" Condition="'$(LocalSystemGlobalizationNative)' != ''"/>
```
SystemGlobalizationNative is internal implementation. We want to user-facing property to be named in terms what users understand and that is serlf-explanatory, like `ICUStaticLinking`. (Feel free to come up with a better name.)

jkotas replied:
More ideas: `StaticICU` , `StaticICULinking`, ... .

jkotas commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:
```diff
@@ -74,8 +76,16 @@ The .NET Foundation licenses this file to you under the MIT license.
+
     <ItemGroup>
       <LinkerArg Include="@(NativeLibrary)" />
+      <LinkerArg Include="@(StaticICULibs)" Condition="'$(LocalSystemGlobalizationNative)' != ''" />
```
I would move this below where we have the other libraries (there is line for `licucore` on Mac).

jkotas commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:
```diff
@@ -74,8 +76,16 @@ The .NET Foundation licenses this file to you under the MIT license.
+    <Message Text="IlcPath = $(IlcPath)" />
+    <Message Text="IlcHostPackagePath = $(IlcHostPackagePath)" />
+
+    <Exec Command="$(IlcPath)/source_code/libs/System.Globalization.Native/build-nativeaot-shim-standalone.sh $(IlcPath)/../ $(IntermediateOutputPath)" Condition="'$(LocalSystemGlobalizationNative)' != '' and '$(IlcPath)' != ''"/>
```
We should use either `IlcHostPackagePath` everywhere or `IlcPath` everywhere. I do not see the reason why we would want to switch back and forth.

Beau-Gosse-dev replied:
@jkotas is `IlcHostPackagePath` still set even if you use `<PublishAot>true</PublishAot>` in your project file and don't reference the ILCompiler NuGet package?

jkotas commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:
```diff
@@ -74,8 +76,16 @@ The .NET Foundation licenses this file to you under the MIT license.
+    <Message Text="IlcPath = $(IlcPath)" />
+    <Message Text="IlcHostPackagePath = $(IlcHostPackagePath)" />
+
+    <Exec Command="$(IlcPath)/source_code/libs/System.Globalization.Native/build-nativeaot-shim-standalone.sh $(IlcPath)/../ $(IntermediateOutputPath)" Condition="'$(LocalSystemGlobalizationNative)' != '' and '$(IlcPath)' != ''"/>
```
You can assume that `IlcPath` / `IlcHostPackagePath` is always set. No need to check it for null.

Commit:
Address feedback from 08/04

* Rename option to ICUStaticLinking
* Remove <Message ..
* Compile without copying sources
* Trim list of defines
* Emit error when run on Linux in CMakeLists.txt
* Move StaticICULibs to near licucore on Mac
* Use only IlcHostPackagePath

smhmhmd committed:
Merge branch 'dotnet:main' into build

Beau-Gosse-dev commented:
I tested as of commit [3fcba3d](https://github.com/dotnet/runtime/pull/72896/commits/3fcba3dfa90bf597fffd1b83f2e85d619fde851b) on ARM by adding the locally built ILCompiler NuGet package to a sample project and it worked well.

jkotas commented:
I will take another look once I come back from vacation.

Commit:
Fix DirectPInvoke for library built by user

smhmhmd committed:
Merge branch 'dotnet:main' into build

Beau-Gosse-dev commented:
Again, I tested this on Linux ARM and it worked with our sample app.

smhmhmd committed:
Merge branch 'dotnet:main' into build

jkotas commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:
```diff
@@ -74,6 +77,8 @@ The .NET Foundation licenses this file to you under the MIT license.
       <NativeFramework Include="GSS" />
     </ItemGroup>
 
+    <Exec Command="$(IlcHostPackagePath)/source_code/libs/System.Globalization.Native/static-build/build-nativeaot-shim-standalone.sh $(IlcHostPackagePath)/ $(IntermediateOutputPath)" Condition="'$(StaticICULinking)' != '' and '$(IlcHostPackagePath)' != ''"/>
```
Do we need IlcHostPackagePath check here? When is this path empty?

jkotas commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:
```diff
@@ -52,7 +52,7 @@ The .NET Foundation licenses this file to you under the MIT license.
     <ItemGroup>
       <NetCoreAppNativeLibrary Include="System.Native" />
-      <NetCoreAppNativeLibrary Include="System.Globalization.Native" />
+      <NetCoreAppNativeLibrary Include="System.Globalization.Native" Condition="'$(StaticICULinking)' == ''" />
```
```suggestion
      <NetCoreAppNativeLibrary Include="System.Globalization.Native" Condition="'$(StaticICULinking)' != 'true'" />
```
And similar for all other checks.

I assume that StaticICULinking property is meant to be set to true/false, and so `<StaticICULinking>false</StaticICULinking>` should keep it turned off.

jkotas commented on src/native/libs/System.Globalization.Native/static-build/CMakeLists.txt:
```diff
@@ -0,0 +1,47 @@
+add_compile_options(-Wno-extra-semi-stmt)
+add_compile_options(-Wno-unknown-warning-option)
+
+add_definitions(-D_GNU_SOURCE -g -fPIC -DHOST_64BIT -DHOST_AMD64 -DHOST_UNIX -DPALEXPORT=EXTERN_C -DTARGET_64BIT -DTARGET_AMD64 -DTARGET_LINUX -DTARGET_UNIX -D_FILE_OFFSET_BITS=64 -DSTATIC_ICU -DSTATIC_SHIM_COMPILE)
```
`-g` and `-fPIC` are not definitions and listed in `target_compile_options`. Do we need them here?

jkotas replied:
`-DHOST_AMD64` is x64 specific define. This should be either deleted or replaced with conditional arch-specific define.

jkotas commented on src/native/libs/System.Globalization.Native/static-build/CMakeLists.txt:
```diff
@@ -0,0 +1,47 @@
+    ${NATIVEGLOBALIZATION_SOURCES}
+)
+
+add_compile_options(-Wno-switch-enum)
+add_compile_options(-Wno-covered-switch-default)
+
+# Workaround for warnings produced by ICU headers
+add_compile_options(-Wno-reserved-id-macro)
+add_compile_options(-Wno-documentation)
+add_compile_options(-Wno-documentation-unknown-command)
+
+# Workaround for https://unicode-org.atlassian.net/browse/ICU-20601
+add_compile_options(-Wno-extra-semi-stmt)
+add_compile_options(-Wno-unknown-warning-option)
```
```suggestion
add_compile_options(-Wno-switch-enum)
add_compile_options(-Wno-covered-switch-default)

# Workaround for warnings produced by ICU headers
add_compile_options(-Wno-reserved-id-macro)
add_compile_options(-Wno-documentation)
add_compile_options(-Wno-documentation-unknown-command)

# Workaround for https://unicode-org.atlassian.net/browse/ICU-20601
add_compile_options(-Wno-extra-semi-stmt)
add_compile_options(-Wno-unknown-warning-option)
```
I would delete this whole thing and instead compile with minimal warning level. It is what we do in dotnet/runtime release braches.

The warnings in C/C++ are somewhat useful for day-to-day development, but they are pain to deal with in released bits where it is hard to just fix them.

jkotas commented on src/native/libs/System.Globalization.Native/static-build/CMakeLists.txt:
```diff
@@ -0,0 +1,47 @@
+
+add_definitions(-D_GNU_SOURCE -g -fPIC -DHOST_64BIT -DHOST_AMD64 -DHOST_UNIX -DPALEXPORT=EXTERN_C -DTARGET_64BIT -DTARGET_AMD64 -DTARGET_LINUX -DTARGET_UNIX -D_FILE_OFFSET_BITS=64 -DSTATIC_ICU -DSTATIC_SHIM_COMPILE)
+
+target_compile_options(System.Globalization.Native PUBLIC -std=gnu99 -Wno-declaration-after-statement -D_GNU_SOURCE -g -fPIC -O0 -g -Wall -fno-omit-frame-pointer -fms-extensions -fwrapv -fstack-protector-strong)
```
> `-O0`

Why are you disabling optimizations?

jkotas replied:
What is the minimal set of options required here to make things compile? It may be best to make this as minimal as possible.

jkotas commented on src/native/libs/System.Globalization.Native/static-build/CMakeLists.txt:
```diff
@@ -0,0 +1,47 @@
+cmake_minimum_required(VERSION 3.6)
+
+project(System.Globalization.Native.User.Machine.Build C)
```
The PR seems to be using multiple different names for the same thing:

- local build (PR title)
- static build (directory name)
- user machine build (this project name)

It would be nice to standardize on a single name. I like "local build" the best out of these options.

jkotas commented on src/native/libs/System.Globalization.Native/static-build/build-nativeaot-shim-standalone.sh:
```diff
@@ -0,0 +1,26 @@
+#!/bin/sh
```
Nit: This script can be called just `build.sh`. The directory name is  enough to disambiguate it from other `build.sh` scripts.

jkotas commented on src/installer/pkg/projects/Microsoft.DotNet.ILCompiler/Microsoft.DotNet.ILCompiler.pkgproj:
```diff
@@ -21,6 +21,10 @@
       <File Include="$(CoreCLRBuildIntegrationDir)*" TargetPath="build" />
       <File Include="$(CoreCLRILCompilerDir)netstandard\*" TargetPath="tools/netstandard" />
       <File Include="sdk\Sdk.props" TargetPath="Sdk" />
+      <File Include="$(MSBuildThisFileDirectory)\..\..\..\..\native\libs\System.Globalization.Native\*" TargetPath="source_code/libs/System.Globalization.Native"/>
```
There is actually an existing directory structure convention for native source. Could you please rename `source_code` to `native/src` ?

For example, https://nuget.info/packages/Microsoft.NETCore.App.Runtime.browser-wasm/6.0.0-preview.3.21201.4 package has `native/src` directory like that.

Commit:
Feedback on 08/25: rename static-build, "source_code" and cleanup compile defs

* Remove IlcHostPackagePath check
* StaticICULinking property is true/false
* Remove -g and -fPIC from definitions
* Use minimal warning level
* Remove -DHOST_AMD64
* Remove -O0
* Remove -fms-extensions -fwrapv
* Standardize on "local build"
* Rename build-nativeaot-shim-standalone.sh to build.sh
* Script name in the error message should match the actual script name
* Rename source_code to native/src

Test code:
$ cat Program.cs
using System.Globalization;

namespace CurrentCulture
{
	class Program
	{
		static void Main(string[] args)
		{
			double val = 1235.56;

			Console.WriteLine($"Current culture: {CultureInfo.CurrentCulture.Name}");
			Console.WriteLine(val);

			CultureInfo.DefaultThreadCurrentCulture = new CultureInfo("en-US");

			Console.WriteLine($"Current culture: {CultureInfo.CurrentCulture.Name}");
			Console.WriteLine(val);
		}
	}
}

$ cat hello-4.csproj
<Project Sdk="Microsoft.NET.Sdk">
 <PropertyGroup>
  <OutputType>Exe</OutputType>
  <TargetFramework>net7.0</TargetFramework>
  <RootNamespace>hello_4</RootNamespace>
  <ImplicitUsings>enable</ImplicitUsings>
  <Nullable>enable</Nullable>
 </PropertyGroup>

$ cat nuget.config
<?xml version="1.0" encoding="utf-8"?>
<configuration>
 <packageSources>
  <!--To inherit the global NuGet package sources remove the <clear/> line below -->
  <clear />
    <add key="repositoryPath" value="/home/ubuntu/runtime/artifacts/packages/Debug/Shipping" />
    <add key="dotnet7" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet7/nuget/v3/index.json" />
 </packageSources>
</configuration>

Test code build:
rm ./bin/Debug/net7.0/linux-arm64/native/hello-4
~/dotnet-sdk/dotnet clean
~/dotnet-sdk/dotnet publish /p:PublishAot=true /p:StaticICULinking=true -bl -r linux-arm64 --self-contained -v d 2>&1 > log
./bin/Debug/net7.0/linux-arm64/native/hello-4

Tested on x86-64 and arm64

smhmhmd committed:
Merge branch 'dotnet:main' into build

Beau-Gosse-dev commented:
@smhmhmd and I both tested the recent commit across ARM64 and x64 with the below sample code, and were able to see that ulocdata_getCLDRVersion_66 was statically linked in the binary. As expected the resulting binary is much bigger with this new static linking. Up from 20MB unstripped to 50MB unstripped.

Published with `~/dotnet-sdk/dotnet publish /p:PublishAot=true /p:StaticICULinking=true -bl -r linux-arm64 --self-contained`

```
namespace CurrentCulture
{
        class Program
        {
                static void Main(string[] args)
                {
                        double val = 1235.56;

                        Console.WriteLine($"Current culture: {CultureInfo.CurrentCulture.Name}");
                        Console.WriteLine(val);

                        CultureInfo.DefaultThreadCurrentCulture = new CultureInfo("en-US");

                        Console.WriteLine($"Current culture: {CultureInfo.CurrentCulture.Name}");
                        Console.WriteLine(val);
                }
        }
}
```

smhmhmd committed:
Merge branch 'dotnet:main' into build

smhmhmd committed:
Merge branch 'dotnet:main' into build

smhmhmd committed:
Merge branch 'dotnet:main' into build

jkotas commented on src/installer/pkg/projects/Microsoft.DotNet.ILCompiler/Microsoft.DotNet.ILCompiler.pkgproj:
```diff
@@ -21,6 +21,10 @@
       <File Include="$(CoreCLRBuildIntegrationDir)*" TargetPath="build" />
       <File Include="$(CoreCLRILCompilerDir)netstandard\*" TargetPath="tools/netstandard" />
       <File Include="sdk\Sdk.props" TargetPath="Sdk" />
+      <File Include="$(MSBuildThisFileDirectory)\..\..\..\..\native\libs\System.Globalization.Native\*" TargetPath="native/src/libs/System.Globalization.Native"/>
```
These sources should be only in the previous item-group so that they are only included in the RID-specific package (e.g. runtime.linux-x64.microsoft.dotnet.ilcompiler). We do not need these sources in the entrypoint package (e.g. microsoft.dotnet.ilcompiler) - they won't be ever used from the entrypoint package.

smhmhmd replied:
>These sources should be only in the previous item-group

I tried that but the user build does not pick up runtime.linux-x64.microsoft.dotnet.ilcompiler as a package reference in the csproj.
I will try some more, if I am still blocked, I will update PR with the change that conditionalizes the parent CMakeLists.txt.

smhmhmd replied:
>Where is the runtime.linux-x64.microsoft.dotnet.ilcompiler package not getting picked up for you?

* The same steps are able to build the user binary without error by using this:
`<PackageReference Include="Microsoft.Dotnet.ILCompiler" Version="8.0.0-dev" /> 
`
* If I remove the PackageReference line, I get the same error as below.

```
ubuntu@ip-10-0-0-223:~/hello$ cat hello.csproj
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net7.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <PublishAot>true</PublishAot>
  </PropertyGroup>

  <ItemGroup>
     <PackageReference Include="runtime.linux-arm64.Microsoft.DotNet.ILCompiler" Version="8.0.0-dev" />
  </ItemGroup>
</Project>

ubuntu@ip-10-0-0-223:~/hello$ cat build.sh 
\rm -rf bin obj
~/dotnet-sdk/dotnet publish -c Release -r linux-arm64 /p:PublishAot=true /p:StaticICULinking=true -bl --self-contained -v d 2>&1 > log
ubuntu@ip-10-0-0-223:~/hello$ sh build.sh 
ubuntu@ip-10-0-0-223:~/hello$ tail log

       "/home/ubuntu/hello/hello.csproj" (Restore target) (1) ->
       (Restore target) -> 
         /home/ubuntu/hello/hello.csproj : error NU1102: Unable to find package runtime.linux-arm64.Microsoft.DotNet.ILCompiler with version (= 7.0.0-rc.1.22423.16)
       /home/ubuntu/hello/hello.csproj : error NU1102:   - Found 1 version(s) in repositoryPath [ Nearest version: 8.0.0-dev ]

    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.15

ubuntu@ip-10-0-0-223:~/hello$ cat nuget.config
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <!--To inherit the global NuGet package sources remove the <clear/> line below -->
    <clear />
    <add key="repositoryPath" value="/home/ubuntu/runtime/artifacts/packages/Debug/Shipping"/>
  </packageSources>
</configuration>

```

jkotas replied:
You have to use the explicit package reference for the `-dev` package that you built locally. it overrides the reference to the shipping package version that is hardcoded in SDK.

smhmhmd replied:
>You have to use the explicit package reference for the -dev package that you built locally. it overrides the reference to the shipping package version that is hardcoded in SDK.

I have the explicit package reference as:
` <PackageReference Include="runtime.linux-arm64.Microsoft.DotNet.ILCompiler" Version="8.0.0-dev" />`

```
ubuntu@ip-10-0-0-223:~/hello$ cat hello.csproj
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net7.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <PublishAot>true</PublishAot>
  </PropertyGroup>

  <ItemGroup>
     <PackageReference Include="runtime.linux-arm64.Microsoft.DotNet.ILCompiler" Version="8.0.0-dev" />
  </ItemGroup>
</Project>

```

jkotas replied:
Our docs typically reference `Microsoft.DotNet.ILCompiler` and let the build figure out the RID-specific name. Referencing the RID-specific package directly works too. Is there something that does not work for you if you just reference `Microsoft.DotNet.ILCompiler` ?

Either way, I still do not see the problem with moving this to the RID-specific ItemGroup.

smhmhmd replied:
>Either way, I still do not see the problem with moving this to the RID-specific ItemGroup.

Yes, I moved the to the previous itemGroup and it worked fine, I will update the change

smhmhmd commented on src/native/libs/System.Globalization.Native/local-build/CMakeLists.txt:
```diff
@@ -0,0 +1,35 @@
+
+if(CMAKE_SYSTEM_NAME STREQUAL "Linux") # Only Linux for now
+
+configure_file(../config.h.in config.h)
```
Good catch, `HAVE_UDAT_STANDALONE_SHORTER_WEEKDAYS` also needs to be 1
The issue is that the CMakeLists.txt in the upper directory has `include(configure.cmake)`, which uses `CMAKE_CURRENT_SOURCE_DIR` and `CMAKE_CURRENT_BINARY_DIR` which cannot be modified and I didnt want to touch the parent CMakefiles.txt, so, I had replaced with `configure_file`.  I should reuse configure_cmake without duplicating it.

Would it be okay to change the parent CMakeLists.txt:

from:
```
configure_file(
    ${CMAKE_CURRENT_SOURCE_DIR}/config.h.in
    ${CMAKE_CURRENT_BINARY_DIR}/config.h)
```
to 
```
if(NOT DEFINED CMAKE_SOURCE_DIR_VAR)
   set(CMAKE_SOURCE_DIR_VAR, ${CMAKE_CURRENT_SOURCE_DIR})
endif()

if(NOT DEFINED CMAKE_BINARY_DIR_VAR)
   setCMAKE_BINARY_DIR_VAR, ${CMAKE_CURRENT_BINARY_DIR})
endif()

configure_file(
    ${CMAKE_SOURCE_DIR_VAR}/config.h.in
    ${CMAKE_BINARY_DIR_VAR}/config.h)
```
With above change, local-build/build.sh can invoke cmake with
  `-DCMAKE_SOURCE_DIR_VAR="../"` and `-DCMAKE_BINARY_DIR_VAR="../"`

smhmhmd replied:
Also, can I set `CLR_CMAKE_TARGET_UNIX=1` or do I need to include configureplatform.cmake

janvorli replied:
It seems unnecessarily complicated, let me take a look at the build changes to understand why we need to have the config.h in a different directory.

janvorli replied:
I wonder - wouldn't it be simpler to modify the creation of the package so that we copy everything from the original (src/native/libs/System.Globalization.Native) directory except of the CMakeLists.txt and then copy this new CMakeLists.txt there too (to the same directory)? That way, you would not have problems with the configuration.
Alternatively, instead of having a separate CMakeLists.txt for the local build, you could reuse the original one and just conditionalize it based on a cmake variable (you could use the STATIC_SHIM_COMPILE that you already have)  that we would pass to cmake using the -D option in the new build.sh. That way you wouldn't even need to duplicate the list of source files.

smhmhmd replied:
>you could reuse the original one and just conditionalize it based on a cmake variable
Sounds good, I will try that, sounds like Jan is onboard as well.
CLR_CMAKE_TARGET_UNIX might still need to be set.

smhmhmd replied:
>I wonder - wouldn't it be simpler to modify the creation of the package so that we copy everything 
Jan wanted to avoid copying in an earlier patch

jkotas replied:
I wanted avoid copying at build time on the target machine. I would be fine with doing the renaming when we create the package.

> > you could reuse the original one and just conditionalize it based on a cmake variable

> Sounds good, I will try that, sounds like Jan is onboard as well.

Yes, I like this option the most. It is the most sustainable option long term, especially if this gets expanded to other native libraries.

smhmhmd replied:
>Yes, I like this option the most. It is the most sustainable option long term, especially if this gets expanded to other native libraries.

Thanks, incorporated change to use only `src/native/libs/System.Globalization.Native/CMakeLists.txt`
Conditionalized above CMakeLists.txt for local build, moved the build.sh into parent as local_build.sh and removed local-build directory.

```
ubuntu@ip-10-0-0-223:~/hello$ cat ./obj/Debug/net7.0/linux-arm64/libs/System.Globalization.Native/build/config.h
#pragma once

#define HAVE_UDAT_STANDALONE_SHORTER_WEEKDAYS 1
#define HAVE_SET_MAX_VARIABLE 1
```

Commit:
Feedback on 09/02

*  Incorporated change to use only src/native/libs/System.Globalization.Native/CMakeLists.txt
*  Conditionalized CMakeLists.txt for local build
*  Moved the build.sh into parent as local_build.sh and removed local-build directory.
*  Rename STATIC_SHIM_COMPILE to LOCAL_BUILD

smhmhmd committed:
Move file includes to previous itemgroup

Move the includes in Microsoft.DotNet.ILCompiler.pkgproj to
<ItemGroup Condition="'$(PackageTargetRuntime)' != ''">

Commit:
Move file includes to previous itemgroup

Move the includes in Microsoft.DotNet.ILCompiler.pkgproj to
<ItemGroup Condition="'$(PackageTargetRuntime)' != ''">

smhmhmd committed:
Merge branch 'build' of https://github.com/smhmhmd/runtime into build

smhmhmd committed:
Merge branch 'dotnet:main' into build

jkotas commented:
One last thing: Could you please add a section "Using statically linked ICU" to https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/docs/compiling.md that documents this feature? (what it does and why one may want to use it, additional pre-requisites to install in addition to https://docs.microsoft.com/en-us/dotnet/core/deploying/native-aot/#prerequisites, etc.)

smhmhmd commented:
> Could you please add a section "Using statically linked ICU" to https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/docs/compiling.md that documents this feature 

Thanks, I will add in next patch

smhmhmd committed:
Update section in compiling.md

* Add "Using statically linked ICU" to compiling.md
* Remove define check for LOCAL_BUILD in CMakeLists.txt

jkotas commented on src/coreclr/nativeaot/docs/compiling.md:
```diff
@@ -68,3 +68,22 @@ You also need to specify the sysroot directory for Clang using the `SysRoot` pro
 You may also follow [cross-building instructions](../../../../docs/workflow/building/coreclr/cross-building.md) to create your own sysroot directory.
+
+## Using statically linked ICU
+This feature can statically link libicu libraries (such as libicui18n.a) into your applications at build time. <br>
```
```suggestion
This feature can statically link libicu libraries (such as libicui18n.a) into your applications at build time.
```

jkotas committed:
Update src/coreclr/nativeaot/docs/compiling.md

jkotas commented on src/coreclr/nativeaot/docs/compiling.md:
````diff
@@ -68,3 +68,22 @@ You also need to specify the sysroot directory for Clang using the `SysRoot` pro
+    <PropertyGroup>
+      <StaticICULinking>true</StaticICULinking>
+    </PropertyGroup>
+```
````
```suggestion
```

This feature is only supported on Linux.
```

jkotas committed:
Update src/coreclr/nativeaot/docs/compiling.md

jkotas commented on src/coreclr/nativeaot/docs/compiling.md:
````diff
@@ -68,3 +68,21 @@ You also need to specify the sysroot directory for Clang using the `SysRoot` pro
+```xml
+    <PropertyGroup>
+      <StaticICULinking>true</StaticICULinking>
+    </PropertyGroup>
````
```suggestion
    </PropertyGroup>
```

This feature is only supported on Linux.
```

jkotas committed:
Update src/coreclr/nativeaot/docs/compiling.md

jkotas reviewed (approved):
LGTM! Thank you for working on this!

@dotnet/ilc-contrib @dotnet/area-system-globalization Any additional feedback?

MichalStrehovsky reviewed (approved):
Looks great otherwise.

If you would like to make sure we're testing this configuration and it doesn't regress, add a test under src/tests/nativeaot/SmokeTests.

smhmhmd committed:
Update src/coreclr/nativeaot/docs/compiling.md

Co-authored-by: Michal Strehovský <MichalStrehovsky@users.noreply.github.com>

smhmhmd commented:
> If you would like to make sure we're testing this configuration and it doesn't regress, add a test under src/tests/nativeaot/SmokeTests.

Thanks, I will add a test under src/tests/nativeaot/SmokeTests

smhmhmd committed:
Update src/installer/pkg/projects/Microsoft.DotNet.ILCompiler/Microsoft.DotNet.ILCompiler.pkgproj

Co-authored-by: Michal Strehovský <MichalStrehovsky@users.noreply.github.com>

smhmhmd committed:
Merge branch 'dotnet:main' into build
